bundles/org.openhab.binding.vallox/src/main/java/org/openhab/binding/vallox/internal/se/connection/ValloxIpConnector.java

```diff
@@ -0,0 +1,133 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ValloxIpConnector} is creates TCP/IP connection to Vallox.
+ *
+ * @author Miika Jukka - Initial contribution
+ */
+@NonNullByDefault
+public class ValloxIpConnector extends ValloxBaseConnector {
+
+    private final Logger logger = LoggerFactory.getLogger(ValloxIpConnector.class);
+
+    private @Nullable Thread readerThread;
+    private Socket socket = new Socket();
+
+    public ValloxIpConnector(ScheduledExecutorService scheduler) {
+        super(null, scheduler);
+        logger.debug("Tcp Connection initialized");
+    }
+
+    /**
+     * Connect to socket
+     */
+    @Override
+    public void connect(ValloxSEConfiguration config) throws IOException {
+        if (isConnected()) {
+            return;
+        }
+        buffer.clear();
+        panelNumber = config.getPanelAsByte();
+        socket = new Socket();
+        socket.connect(new InetSocketAddress(config.tcpHost, config.tcpPort), ValloxSEConstants.CONNECTION_TIMEOUT);
+        socket.setSoTimeout(ValloxSEConstants.SOCKET_READ_TIMEOUT);
+        inputStream = socket.getInputStream();
+        outputStream = socket.getOutputStream();
+        logger.debug("Connected to {}:{}", config.tcpHost, config.tcpPort);
+
+        readerThread = new TelegramReader();
+        readerThread.start();
+        connected = true;
+    }
+
+    /**
+     * Close socket
+     */
+    @SuppressWarnings("null") // readerThread .interrupt() and .join()
+    @Override
+    public void close() {
+        super.close();
+        if (readerThread != null) {
+            logger.debug("Interrupting telegram listener");
+            readerThread.interrupt();
+            try {
+                readerThread.join();
+            } catch (InterruptedException e) {
+                // Do nothing
+            }
+        }
+        try {
+            socket.close();
+        } catch (Exception e) {
+            logger.debug("Exception closing connection: ", e);
+        }
+        readerThread = null;
+        connected = false;
+        logger.debug("Closed");
+    }
+
+    /**
+     * {@link Thread} implementation for reading telegrams
+     *
+     * @author Miika Jukka
+     */
+    @SuppressWarnings("null") // inputStream .available() and .read()
+    @NonNullByDefault
+    private class TelegramReader extends Thread {
+        boolean interrupted = false;
+
+        @Override
+        public void interrupt() {
+            interrupted = true;
+            super.interrupt();
+        }
+
+        @Override
+        public void run() {
+            logger.debug("Data listener started");
+            while (!interrupted && inputStream != null) {
+                try {
+                    while (inputStream.available() > 0) {
+                        buffer.add((byte) inputStream.read());
+                    }
+                    handleBuffer();
+                    Thread.sleep(200); // Avoid high CPU usage by sleeping between loops
```

Comment:
Instead of sleeping you should be relying on `InputStream.read` blocking until data becomes available. `read` will return -1 when it reaches the end of the stream.

Reply:
Done.